src/common/mailing.ts

```diff
@@ -171,3 +179,134 @@ export const addPrivateSourceJoinParams = ({
 
   return urlObj.toString();
 };
+
+const ITEMS_PER_DESTROY = 4000;
+const ITEMS_PER_IDENTIFY = 250;
+
+interface SyncSubscriptionsWithActiveStateProps {
+  con: DataSource;
+  users: GetUsersActiveState;
+}
+
+export const syncSubscriptionsWithActiveState = async ({
+  con,
+  users: { inactiveUsers, downgradeUsers, reactivateUsers },
+}: SyncSubscriptionsWithActiveStateProps) => {
+  await blockingBatchRunner({
+    data: reactivateUsers,
+    runner: async (batch) => {
+      const validReactivateUsers = await con.getRepository(User).find({
+        select: ['id'],
+        where: { id: In(batch), cioRegistered: false },
+      });
+
+      if (validReactivateUsers.length === 0) {
+        return true;
+      }
+
+      await blockingBatchRunner({
+        batchLimit: ITEMS_PER_IDENTIFY,
+        data: validReactivateUsers.map(({ id }) => id),
+        runner: async (ids) => {
+          const users = await con
+            .getRepository(User)
+            .find({ where: { id: In(ids) } });
+
+          const data = await Promise.all(
+            users.map((user) =>
+              generateIdentifyObject(con, toChangeObject(user)),
+            ),
+          );
+
+          await callWithRetryDefault({
+            callback: () => cioV2.request.post('/users', { batch: data }),
+            onSuccess: async () => {
+              await con
+                .getRepository(User)
+                .update({ id: In(ids) }, { cioRegistered: true });
+            },
+            onFailure: (err) => {
+              logger.info({ err }, 'Failed to add users to CIO');
+            },
+          });
+
+          await setTimeout(20); // wait for a bit to avoid rate limiting
+        },
+      });
+    },
+  });
+
+  // inactive for 12 weeks: remove from CIO
+  await blockingBatchRunner({
+    data: inactiveUsers,
+    runner: async (batch) => {
+      const validInactiveUsers = await con.getRepository(User).find({
+        select: ['id'],
+        where: { id: In(batch), cioRegistered: true },
+      });
+
+      if (validInactiveUsers.length === 0) {
+        return true;
+      }
+
+      await blockingBatchRunner({
+        batchLimit: ITEMS_PER_DESTROY,
+        data: validInactiveUsers.map(({ id }) => id),
+        runner: async (ids) => {
+          const data = ids.map((id) => ({
+            action: 'destroy',
+            type: 'person',
+            identifiers: { id },
+          }));
+
+          await callWithRetryDefault({
+            callback: () => cioV2.request.post('/users', { batch: data }),
+            onSuccess: async () => {
+              await con.getRepository(User).update(
```

Comment:
We also need to remove the digest entry right?

Reply:
Right. Good catch. Let me add the removal 👀